far/vmenu.cpp

```diff
@@ -140,15 +140,45 @@ void VMenu::ResetCursor()
 	GetCursorType(PrevCursorVisible,PrevCursorSize);
 }
 
+template <typename coll, typename pred>
+int FindNearestSelectableItem(const coll& Coll, const int Pos, pred&& Pred, const bool GoBackward, const bool DoWrap)
+{
+	assert(0 <= Pos && Pos < static_cast<int>(Coll.size()));
+
+	const auto FindPos{
+		[&](const auto First, const auto Second)
+		{
+			const auto FindPos{
+				[&](const auto Part)
+				{
+					const auto Found{ std::ranges::find_if(Part, Pred) };
+					if (Found == std::ranges::end(Part)) return -1;
+					return static_cast<int>(std::ranges::distance(&*std::ranges::begin(Coll), &*Found));
+				} };
+
+			if (const auto Found{ FindPos(Coll | First) }; Found != -1) return Found;
+			if (const auto Found{ FindPos(Coll | Second) }; Found != -1) return Found;
+			return -1;
+		} };
+
+	return GoBackward
```

Comment:
Just to confirm that I understand the algorithm correctly:

```
For a collection ABCDP1234 where P denotes Pos:

 back &  wrap  => try PDCBA then 4321 then give up
 back & !wrap  => try PDCBA then 1234 then give up
!back &  wrap  => try P1234 then ABCD then give up
!back & !wrap  => try P1234 then DCBA then give up
```
?

Reply:
That is correct.

Reply:
Thanks.
I have a subtle feeling that ranges can make it even more abstract-ish.
E.g. if after chopping & optional reversing we apply `| filter(pred)`, then the lookup could be as simple as taking the first item, if any.

Reply:
That's neat! Thank you. I am still learning ranges myself. This is partly the reason why I try to use them whenever reasonable. And based on my experience with C# LINQ, it is reasonable way more often than we tend to think. As an aside, with C++ aggressive inlining and good optimization, the resulting codegen is usually as good as hand-written loops.